In [44]:
#IPython is what you are using now to run the notebook
import IPython
print( "IPython version:      %6.6s (need at least 1.0)" % IPython.__version__)

# Numpy is a library for working with arrays and matrices
import numpy as np
print( "Numpy version:        %6.6s (need at least 1.7.1)" % np.__version__)

# SciPy implements many different numerical algorithms
import scipy as sp
print( "SciPy version:        %6.6s (need at least 0.12.0)" % sp.__version__)

# Pandas makes working with data tables easier
import pandas as pd
print( "Pandas version:       %6.6s (need at least 0.11.0)" % pd.__version__)

# Module for plotting
import matplotlib.pyplot as plt  
from pylab import *
print( "Mapltolib version:    %6.6s (need at least 1.2.1)" %
       matplotlib.__version__)
%matplotlib inline
# necessary for in-line graphics

# SciKit Learn implements several Machine Learning algorithms
import sklearn
print( "Scikit-Learn version: %6.6s (need at least 0.13.1)" %
       sklearn.__version__)
import os
# for certain system-related functions

import seaborn as sns
print( "Seaborn version: %6.6s (need at least 0.13.1)" %
       sns.__version__)

from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC

IPython version:       7.6.1 (need at least 1.0)
Numpy version:        1.16.4 (need at least 1.7.1)
SciPy version:         1.2.1 (need at least 0.12.0)
Pandas version:       0.24.2 (need at least 0.11.0)
Mapltolib version:     3.1.0 (need at least 1.2.1)
Scikit-Learn version: 0.21.2 (need at least 0.13.1)
Seaborn version:  0.9.0 (need at least 0.13.1)


### 1 Explore and prepare the data 

In [45]:
data=pd.read_csv("C:/Users/roshn/Desktop/IMT 574A/wvs.csv",sep='\t')
data=pd.DataFrame(data)

#Checking for the top 5 rows 

print(data.head(5))

#Size of the dataset
print("Number of rows and columns in the data",data.shape)

#Columns in the dataset
print("Columns in the dataset",data.columns)

#description of the dataset
print("Summary statistics of the dataset")
print(data.describe())


   V2  V4  V5  V6  V7  V8  V9  V10  V11  V12  ...  MN_228S8  MN_229A  MN_230A  \
0  12   1   1   1  -2   1   1    2    1    1  ...         3       -3       -3   
1  12   1   2   3   4   2   2    2    2    2  ...         3       -3       -3   
2  12   1   3   2   4   2   1    2    2    2  ...         4        1        1   
3  12   1   1   3   4   3   1    2    1    2  ...         2        2        1   
4  12   1   1   1   2   1   1    1    3    2  ...         2        2        1   

   MN_233A  MN_237B1  MN_249A1  MN_249A3  I_RELIGBEL  I_NORM1  I_VOICE1  
0       -3        -3         1         1         0.0      1.0      0.00  
1       -3        -3         2        -1         0.0      1.0      0.66  
2        2        -3         1         1         0.0      1.0      0.33  
3        2        -3         1         2         0.0      1.0      0.00  
4        2        -3         1         2         0.0      1.0      0.66  

[5 rows x 328 columns]
Number of rows and columns in the data (90350

### 1.1 Number of responses and predictors

In [46]:
print("Number of Responses - ", data.shape[0])
print("Number of variables - ",data.shape[1])

Number of Responses -  90350
Number of variables -  328


### 1.2 Summary statistics for V204 : is abortion justifiable

In [47]:
#Summary statistics of the data column
print("Summary statistics of V204-Abortion")
print(data.V204.describe())


#Count of values in V204 column
print(data.V204.value_counts())



#Count of non-missing values
print("Number of non missing values(positives)")
print(len(data.loc[data.V204 > 0]))


#Count of negative values
print("Number of missing values(negatives)")
print(len(data.loc[data.V204 < 0]))

Summary statistics of V204-Abortion
count    90350.000000
mean         2.946386
std          2.964040
min         -5.000000
25%          1.000000
50%          2.000000
75%          5.000000
max         10.000000
Name: V204, dtype: float64
 1     40227
 5      9580
 2      7896
 3      6294
 4      4497
 6      4395
 10     4067
 7      3493
 8      3397
-1      2017
 9      1896
-4      1523
-2      1045
-5        23
Name: V204, dtype: int64
Number of non missing values(positives)
85742
Number of missing values(negatives)
4608


In [48]:
val1 = data['V204']
list1 = val1[val1>0].unique()

def prob(num):
    positive_resp = 85742
    prob = (len(val1[val1 == num]) / positive_resp)*100
    return prob 
    
prod_1 = prob(1)
for i in list1:
    prob_value = prob(i) 
    print("probability of opinion" , i , round(prob_value,2))

probability of opinion 1 46.92
probability of opinion 5 11.17
probability of opinion 4 5.24
probability of opinion 6 5.13
probability of opinion 2 9.21
probability of opinion 3 7.34
probability of opinion 10 4.74
probability of opinion 7 4.07
probability of opinion 8 3.96
probability of opinion 9 2.21


### 1.3  Now remove missings - remove everything that are not positive integers for V204 and V2 (country).

In [49]:
#Removing the missing data by the first way: everything that are not positive integers for V204 and V2 (country)
#Copy dataset
cdata=data.copy()
cdata.head()

#Dropping values from V204
cdata=cdata[cdata.V204>0]
print("Total number of rows after dropping",len(cdata))

#Checking if the values <0 got dropped from V204 column
len(cdata.loc[cdata.V204 < 0])

#Dropping values from V2
cdata=cdata[cdata.V2>0]


#Checking if the values <0 got dropped from V2 column
len(cdata.loc[cdata.V2 < 0])

print(cdata.shape)


Total number of rows after dropping 85742
(85742, 328)


### 1.3 b)For all other variables, remove the missings in the sense of missing value on computer.

In [52]:
# Drop rows with any NA values in the selected columns only, except V2 and V204
cdata=cdata.dropna()

#print("Number of observations in the cleaned dataset:")
print(cdata.shape)




(79267, 329)


### 1.4:  In order to simplify the analysis below, create a new binary variable abortion:
#### (1 when V204 > 3 ) (0 for other cases)

In [53]:
#Creating new binary variables in V204
cdata['Abortion']=np.where(cdata.V204>3, 1, 0)
cdata.head(3)


V2  V4  V5  V6  V7  V8  V9  V10  V11  V12  ...  MN_229A  MN_230A  MN_233A  \
0  12   1   1   1  -2   1   1    2    1    1  ...       -3       -3       -3   
1  12   1   2   3   4   2   2    2    2    2  ...       -3       -3       -3   
2  12   1   3   2   4   2   1    2    2    2  ...        1        1        2   

   MN_237B1  MN_249A1  MN_249A3  I_RELIGBEL  I_NORM1  I_VOICE1  Abortion  
0        -3         1         1         0.0      1.0      0.00         0  
1        -3         2        -1         0.0      1.0      0.66         0  
2        -3         1         1         0.0      1.0      0.33         0  

[3 rows x 329 columns]

### 1.5  Compute (pearson) correlation table between abortion and all other variables in the data. Present these variables in descending order according to the absolute value of the correlation.

In [74]:
#Dropping the v204 variable from the data
cdata_copy=cdata.copy()
cdata_copy=pd.DataFrame(cdata_copy)
cdata=cdata.drop(['V204'], axis=1)
cdata.head(3)

#Correlation between Abortion and other variables in the data

corrMatrix=cdata[cdata.columns[1:]].corr()['Abortion'][:-1].sort_values(ascending=False)
corrMatrix.head(5)




V205    0.548653
V203    0.485419
V206    0.446394
V207    0.418271
V9      0.314117
Name: Abortion, dtype: float64

### 1.6 Convert Country Code V2 into dummies. First rename V2 to Country

In [ ]:
print("The data contains 57 dummies")

In [76]:
cdata = cdata.rename(columns = {'V2' : 'country'})

cdata.shape
cdata_copy_1 = pd.get_dummies(cdata, columns = ['country'],drop_first=True)
print(cdata_copy_1.shape)
cdata_copy_1 = pd.DataFrame(cdata_copy_1)

(79267, 384)


### 2) Implement Cross-Validation

In [56]:
def cross_validation(model, X,y,k=6):
   
    from sklearn.model_selection import KFold
    from sklearn.svm import SVR
    from sklearn.metrics import accuracy_score
    from sklearn.metrics import f1_score, precision_score, recall_score
    
    #Obtaining the index values of X and y
    X_indices = np.asarray(X.index)
    y_indices = np.asarray(y.index)
    X_set = set(X_indices.tolist())
    
    accuracy1 = []
    score_accuracy = []
    precision = []
    recall = []
    
    for i in range(k):
       
        validation_indices = set(X_indices[i : : 6].tolist())
        train_indices  = X_set.difference(validation_indices) 
        
        validation_X = X.loc[validation_indices]
        validation_y = y.loc[validation_indices]
        
        train_X = X.loc[train_indices]
        train_y = y.loc[train_indices]
        
        model.fit(train_X, train_y)
        
        y_pred = model.predict(validation_X)
        
        accuracy = accuracy_score(validation_y, y_pred)
        accuracy1.append(accuracy) 
        
        
        score = f1_score(validation_y, y_pred)
        score_accuracy.append(score)
        
        precision.append(precision_score(validation_y, y_pred, average="macro"))
        recall.append(recall_score(validation_y, y_pred, average="macro"))
        
    accuracy  =accuracy1
    score_accuracy = score_accuracy
    
      
        
    return ( (round(np.mean(accuracy) ,2 )) , (round(np.mean(score_accuracy) , 2)) , precision , recall )
    
    
    

### 3.1) Implementation of KNN

In [57]:
#Randomly sampling the data
cdata_copy_crossvalidation = cdata_copy.sample(4000)

In [58]:
# Choosing X and y from the sampled data
X = cdata_copy_crossvalidation.loc[:, cdata_copy_crossvalidation.columns != 'Abortion']
y = cdata_copy_crossvalidation.loc[:, cdata_copy_crossvalidation.columns == 'Abortion']


k = 7
knn_classifier = KNeighborsClassifier(n_neighbors = 7)
validation1 , validation2 , validation11 , validation22 = cross_validation(knn_classifier,X,y,k)
print("mean accuracy of Knn is ", validation1)
print("mean F1 score of Knn is ", validation2)


C:\Users\roshn\Anaconda3\lib\site-packages\ipykernel_launcher.py:29: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
C:\Users\roshn\Anaconda3\lib\site-packages\ipykernel_launcher.py:29: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
C:\Users\roshn\Anaconda3\lib\site-packages\ipykernel_launcher.py:29: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
C:\Users\roshn\Anaconda3\lib\site-packages\ipykernel_launcher.py:29: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
C:\Users\roshn\Anaconda3\lib\site-packages\ipykernel_launcher.py:29: DataConversionWarning: A column

mean accuracy of Knn is  0.75
mean F1 score of Knn is  0.64


C:\Users\roshn\Anaconda3\lib\site-packages\ipykernel_launcher.py:29: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


In [59]:
rangek=np.arange(1,6)

kaccuracy = []
kfscore = []

kvalue = []
for k in rangek:
    knn = KNeighborsClassifier(n_neighbors=k)
    accuracy , fscore,precise,recall = cross_validation(knn,X,y,k)
    kvalue.append(k)
    kaccuracy.append(accuracy)
    kfscore.append(fscore)  
cdata_copy = list(zip(kvalue,kaccuracy,kfscore))

knn_accuracy = pd.DataFrame(cdata_copy , columns=['k value ','Accuracy','Fscore']).sort_values('Accuracy',ascending=False)
knn_fscore = pd.DataFrame(cdata_copy , columns=['Value of k ','Accuracy','Fscore']).sort_values('Fscore',ascending=False)
print("Best value of k with highest accuracy is : ", " " ,knn_accuracy.loc[0,]," Best value of k with highest f score is:",knn_fscore.loc[0,])

#print("Best value of k with highest fscore is : ", " " ,knn_df_fscore.loc[0,])

    

C:\Users\roshn\Anaconda3\lib\site-packages\ipykernel_launcher.py:29: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
C:\Users\roshn\Anaconda3\lib\site-packages\ipykernel_launcher.py:29: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
C:\Users\roshn\Anaconda3\lib\site-packages\ipykernel_launcher.py:29: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
C:\Users\roshn\Anaconda3\lib\site-packages\ipykernel_launcher.py:29: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
C:\Users\roshn\Anaconda3\lib\site-packages\ipykernel_launcher.py:29: DataConversionWarning: A column

Best value of k with highest accuracy is :    k value     1.00
Accuracy    0.72
Fscore      0.62
Name: 0, dtype: float64  Best value of k with highest f score is: Value of k     1.00
Accuracy       0.72
Fscore         0.62
Name: 0, dtype: float64


### 3.2  Logistic Regression

In [60]:
logisticmodel= LogisticRegression()

validation3 , validation4 , valid33 , valid44 = cross_validation(logisticmodel,X,y,k)
print("Logistic Regression has a mean accuracy of :", validation3)
print("Logistic Regression has a mean F1 score of : ", validation4)


C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\linear_

Logistic Regression has a mean accuracy of : 0.81
Logistic Regression has a mean F1 score of :  0.74


### 3.3  Support Vector Machines(SVM):
#### With linear kernel

In [61]:
#Using a linear kernel
svm_model1= SVC(kernel='linear')
validation5 , validation6 , valid55 , valid66 = cross_validation(svm_model1,X,y,k)
print("Support Vector Machine(SVM) model with linear kernel has a mean accuracy of : ", validation5)
print("Support Vector Machine(SVM) model with linear kernel has a mean f1 score of : ", validation6)
print("precision value of SVM : " , valid55[0])
print("recall value of SVM : ", valid66[0])



C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), 

Support Vector Machine(SVM) model with linear kernel has a mean accuracy of :  0.8
Support Vector Machine(SVM) model with linear kernel has a mean f1 score of :  0.72
precision value of SVM :  0.8149749373433584
recall value of SVM :  0.8079398737159551


#### With polynomial kernel

In [39]:
svmmodel2 = SVC(kernel='poly', degree=8)
validation7 , validation8 , valid77 , valid88 = cross_validation(svmmodel2 ,X,y,k)
print("mean accuracy of Support Vector Machine with kernel with polynomial kernel: ", validation7)
print("mean F1 score of Support Vector Machine with kernel with polynomial kernel : ", validation8)



C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to

mean accuracy of Support Vector Machine with kernel with polynomial kernel:  0.99
mean F1 score of Support Vector Machine with kernel with polynomial kernel :  0.99


### 3.4 ) Comparing the models

#### 1.After comparing the models , it was found that the best model is SVM with polynomial kernel with an accuracy of 99%.In terms of F-score too, the best model is SVM with polynomial kernel .
#### Some of the models are really slow like KNN and SVM while it was also observed that logisic regression runs much faster.

#### 2. ) If I have to repeat the exercise with a single model  , I will be picking Logistic Regression, since it has a good accuracy and F1 score and also , the model executes fast

### 4) How large a role does country play?

In [66]:
#with dummy variables 

print("With Dummy Variables")
logisticmodel1 = LogisticRegression()
validation3 , validation4 , val33 , val44 = cross_validation(logisticmodel1,X,y,k)
print("Mean accuracy of Logistic model with dummy variables is:  ", validation3)



With Dummy Variables


C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\linear_

Mean accuracy of Logistic model with dummy variables is:   0.81


In [79]:
#without dummy variables 
# Choosing X and y from the sampled data
X = cdata.loc[:, cdata.columns != 'Abortion']
y = cdata.loc[:, cdata.columns == 'Abortion']


X.drop(["country"], axis=1 , inplace=True)

logisticmodel1 = LogisticRegression()
validation30 , validation40 , val330 , val440 = cross_validation(logisticmodel1,X,y,k)
print("mean accuracy of Logistic Regression is without dummy variables is  ", validation30)



C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\roshn\Anaconda3\lib\site-packages\sklearn\linear_

mean accuracy of Logistic Regression is without dummy variables is   0.83


#### It can be observed that the accuracy increases from 0.81 to 0.83, which is not a very high increase.It can therefore be concluded that country variable doesn't influence our result much,and hence abortion .